worker/tracer/worker.go

```diff
@@ -0,0 +1,211 @@
+				continue
+			}
+
+			req.done <- nil
+
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		}
+	}
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *tracerWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *tracerWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+// GetTracer returns a tracer for the given namespace.
+func (w *tracerWorker) GetTracer(namespace string) (coretracer.Tracer, error) {
+	// First check if we've already got the tracer worker already running. If
+	// we have, then return out quickly. The tracerRunner is the cache, so there
+	// is no need to have a in-memory cache here.
+	if tracer, err := w.workerFromCache(namespace); err != nil {
+		return nil, errors.Trace(err)
+	} else if tracer != nil {
+		return tracer, nil
+	}
+
+	// Enqueue the request as it's either starting up and we need to wait longer
+	// or it's not running and we need to start it.
+	req := traceRequest{
+		namespace: namespace,
+		done:      make(chan error),
+	}
+	select {
+	case w.tracerRequests <- req:
+	case <-w.catacomb.Dying():
+		return nil, w.catacomb.ErrDying()
+	}
+
+	// Wait for the worker loop to indicate it's done.
+	select {
+	case err := <-req.done:
+		// If we know we've got an error, just return that error before
+		// attempting to ask the tracerRunnerWorker.
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	case <-w.catacomb.Dying():
+		return nil, w.catacomb.ErrDying()
+	}
+
+	// This will return a not found error if the request was not honoured.
+	// The error will be logged - we don't crash this worker for bad calls.
+	tracked, err := w.tracerRunner.Worker(namespace, w.catacomb.Dying())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return tracked.(coretracer.Tracer), nil
+}
+
+func (w *tracerWorker) workerFromCache(namespace string) (coretracer.Tracer, error) {
+	// If the worker already exists, return the existing worker early.
+	if tracer, err := w.tracerRunner.Worker(namespace, w.catacomb.Dying()); err == nil {
+		return tracer.(coretracer.Tracer), nil
+	} else if errors.Is(errors.Cause(err), worker.ErrDead) {
+		// Handle the case where the DB runner is dead due to this worker dying.
+		select {
+		case <-w.catacomb.Dying():
+			return nil, w.catacomb.ErrDying()
+		default:
+			return nil, errors.Trace(err)
+		}
+	} else if !errors.Is(errors.Cause(err), errors.NotFound) {
+		// If it's not a NotFound error, return the underlying error. We should
+		// only start a worker if it doesn't exist yet.
+		return nil, errors.Trace(err)
+	}
+	// We didn't find the worker, so return nil, we'll create it in the next
+	// step.
+	return nil, nil
+}
+
+func (w *tracerWorker) initTracer(namespace string) error {
+	err := w.tracerRunner.StartWorker(namespace, func() (worker.Worker, error) {
+		ctx, cancel := w.scopedContext()
+		defer cancel()
+
+		return w.cfg.NewTracerWorker(ctx, namespace, w.cfg.Logger)
```

Comment:
If the tracer isn't enabled per model, we should return a noop version. That way it's hidden away transparently from all the call sites.